decidim/decidim
# Webpacker: Use local NPM installer instead of duplicating the packages folder

#### :tophat: What? Why?
As discussed with @leio10 at https://github.com/decidim/decidim/pull/8136#issuecomment-862317539, we should migrate to using a local NPM package installer instead of duplicating the packages folder in every app.

This is a workaround for installing the linked package dependencies due to a bug in NPM v7 (https://github.com/npm/cli/issues/2339). Until the issue is fixed at NPM, we need this workaround.

The local package installer is a separate NPM package implemented at:
https://github.com/decidim/decidim-npm-local

This tool does not have to follow the Decidim version numbering, it is designed to work on its own. It could be released already to NPM even before there is an actual Decidim release available.

The workaround works as follows:
- It fetches the `decidim` gem installation path using `bundle show decidim`
  * If it finds one, it copies the `packages` folder to the application's `tmp` folder which doesn't go to git
  * If it fails to find the Decidim installation path, it will default to installing the Decidim NPM dependencies from Decidim's git repository using [`degit`](https://www.npmjs.com/package/degit)
- It [packs](https://docs.npmjs.com/cli/v7/commands/npm-pack) all the NPM dependencies to their own TGZ pack files
- It installs the packed dependencies from the TGZ packages which also installs the "linked" dependencies

Benefits compared to the current solution:
- The `packages` folder does not need to be duplicated to the application itself for local and git installations
- The `packages` folder does not need to be duplicated to the design app

If the [related NPM bug](https://github.com/npm/cli/issues/2339) is fixed at some point, we can consider removing this workaround and installing the Decidim NPM dependencies directly using file paths pointing to the `decidim` gem's installation folder (when available, note that there are also installations that do not install the `decidim` main gem but only e.g. `decidim-core`, `decidim-admin`, etc.).

#### :pushpin: Related Issues
- Related to https://github.com/decidim/decidim/pull/8136#issuecomment-862317539 #8119 https://github.com/npm/cli/issues/2339

#### Testing
- Create the development app
- Go to the development app
- Remove the `tmp/npmbuild` folder
- Remove the `node_modules` folder
- Run `npm i` inside the development app
- Start the webpack dev server by running `./bin/webpack-dev-server`
- See that there are no problems running the server and the npm packs are always regenerated when running `npm i` causing no changes to e.g. deployment scripts

#### :clipboard: Checklist
